client/components/web-preview/content.jsx

```diff
@@ -232,7 +231,7 @@ export class WebPreviewContent extends Component {
 		} else {
 			debug( 'preview loaded for url:', this.state.iframeUrl );
 		}
-		if ( this.checkForIframeLoadFailure( caller ) ) {
+		if ( ! this.props.disableRedirects && this.checkForIframeLoadFailure( caller ) ) {
```

Comment:
Cool! I don't notice we have `jetpack-iframe-embed.js`. I think the problem is that I use the API endpoint as the iframe url directly, so it won't have that script

Reply:
The `WebPreview` component is designed to embed documents that have the `jetpack-iframe-embed` script and are able to talk to the parent frame using the specified message protocol. If you don't want that, you can load the document into your own `iframe`, without using `WebPreview` at all. Or does `WebPreview` have any functionality of does it provide any services that you really need here?

Looking at the `/template/demo` REST endpoint, I see that it pipes through a rendered page of that particular theme's demo site. So it actually _is_ a preview of a WordPress site and it would be nice to have the `jetpack-iframe-embed` script active it it. To achieve that, you need to have some query args in the site URL, like:
```
https://rockfieldstarter.wordpress.com?iframe=true&theme_preview=true&frame-nonce=...
```
The `iframe` query arg, together with either `preview` or `theme_preview` query arg, is what triggers the `jetpack-iframe-embed` script. See the `Jetpack_Iframe_Embed::is_embedding_in_iframe` PHP method.

To pass through the `iframe` query arg, you'll need to modify the following statement in the `WPCOM_JSON_API_Template_Demo_Endpoint` PHP class:
```php
$demo_content_response = $template_helpers->get_demo_content(
  $theme_demo_site_id,
  [
    'theme_preview' => 'true',
    'language'      => $language,
  ],
);
```
and pass through also the `iframe` query arg of the REST endpoint, just like it already passes through the `language` param.

You'll probably also need to work with the `iframe-nonce` param, but I don't know the details.

Doing all this, you should arrive at perfect theme demo site previews 🙂

Reply:
Interesting! I've modified `WPCOM_JSON_API_Template_Demo_Endpoint` PHP class and let it support `iframe` query arg here: D67443-code